### ISSUE TITLE
Add a "message_key" to backdrop_set_message() for easily changing, or removing messages.

### ISSUE THREAD

**Issue by MrHaroldA** [backdrop/backdrop-issues#1062]
If you don't want to show a core, or contrib message to the end user, you now have to iterate `$_SESSION['messages']` and search for the (translated, token replaced) string to unset it.

If a message has a unique key, you can easily override a message, or unset it.

`function backdrop_set_message($message = NULL, $type = 'status', $repeat = TRUE, $message_key = FALSE) {`

A simple `unset($_SESSION['messages']['status']['unique_key'])` could remove the message, and `$_SESSION['messages']['status']['unique_key'] = t('Something else');` changes it.

**Comment by MrHaroldA** [comment]
Mentioning #1060 as it fixes a lot of core submit messages.

**Comment by docwilmot** [comment]
A unique key would be a great idea, but not sure about an optional parameter. If you wanted to unset a particular message, it would then be highly unlikely that the caller would have set this key. Everyone would just have continued to write `backdrop_set_message('my_message_string')` anyway.

You could also port https://www.drupal.org/project/disable_messages. :smile:

**Comment by MrHaroldA** [comment]
> not sure about an optional parameter

I'd rather have it mandatory too, but that will break core and (probably) all contrib modules.

Coding something like disabled_messages would be a breeze if all messages have unique keys (like emails have).

**Comment by MrHaroldA** [comment]
How about a fallback on the caller function (using `debug_backtrace()`)?

> limit
As of 5.4.0, this parameter can be used to limit the number of stack frames returned.

`$bt = debug_backtrace(DEBUG_BACKTRACE_IGNORE_ARGS, 2);
$message_key = $bt[1]['function']`

http://php.net/manual/en/function.debug-backtrace.php

**Comment by docwilmot** [comment]
So something like:

```
function backdrop_set_message($message = NULL, $type = 'status', $repeat = TRUE) {
  $bt = debug_backtrace(DEBUG_BACKTRACE_IGNORE_ARGS, 2);
  $message_key = $bt[1]['function'];
  ...
    if ($repeat || !in_array($message, $_SESSION['messages'][$type])) {
      $_SESSION['messages'][$type][$message_key] = $message;
    }
```

But then thats still an API change if `$mess` is now an associative array, not numeric as expected. And if the same function sets two messages, we're in trouble too.

**Comment by MrHaroldA** [comment]
True, true ...

**Comment by klonos** [comment]
It would be amazing if we could solve this. It's always been a pain trying to hide certain messages from the end users based for example on user role. For historic reference, I'll mention [Allow drupal_set_message to store information that can help define visibility (or not). [#1149006]](https://www.drupal.org/node/1149006#comment-10138474) (against core - currently pushed for D9) AS WELL AS [Filter messages by module [#1137722]](https://www.drupal.org/node/1137722#comment-10138700) against [Disable Messages](https://www.drupal.org/project/disable_messages).

**Comment by klonos** [comment]
...some more (these are from my "my issues" list from d.org - there surely must be more out there):

[Add permissions to the update.module to hide warnings [#332796]](https://www.drupal.org/node/332796)
[Don't show aggregator messages to anonymous users [#371157]](https://www.drupal.org/node/371157)

[Suppress status messages to anonymous users when using poormanscron [#867054]](https://www.drupal.org/node/867054)
[Suppress status messages to anonymous users [#902368]](https://www.drupal.org/node/902368)

**Comment by docwilmot** [comment]
I think it's a worthwhile API change though. 2.0?

**Comment by klonos** [comment]
Yep, I think so too.

**Comment by quicksketch** [comment]
I also think this would be a nice-to-have. Although the backtrace idea is clever, I don't think I'd want to depend upon it for identification of messages. Just by moving a message from one function to another would suddenly result in "API changes". If the parameter remains optional, I think anything beyond putting it in an "uncategorized" group or something like that would be presumptuous.

**Comment by docwilmot** [comment]
How about adding a new function: 
```
function backdrop_message( $key, $message = NULL, $type = 'status', $repeat = TRUE)
```

Then change all core `backdrop_set_message()` to that? But then we keep `backdrop_set_message()` as a deprecated function:

```
function backdrop_set_message($message = NULL, $type = 'status', $repeat = TRUE) {
  backdrop_message('legacy', $message, $type', $repeat);
}
```

**Comment by MrHaroldA** [comment]
Cool idea! But should $key be unique? Or is $message stored in a $key-ed array?

**Comment by quicksketch** [comment]
How would we deal with strings that have replacements in them? e.g. `t('Thank you @name for your order')`. If you're going to be able to alter messages, wouldn't you need the replacements to be separated from the string? Then this starts to look like the `watchdog()` function:

```
backdrop_message($key, $message, $replacements, $type);
```

But I've never been a fan of the separation from `t()` in the `watchdog()` function. Besides the additional complexity for developers, it doesn't accommodate for `format_plural()` where different versions of the string are needed for singular/plural.

**Comment by docwilmot** [comment]
So to get what you're saying, if we called `backdrop_message('whatever', t('Thank you @name for your order', ('@name'=> $name)))`:

```
function backdrop_message($message = NULL, $type = 'status', $repeat = TRUE) {
...
    if ($repeat || !in_array($message, $_SESSION['messages'][$type])) {
      $_SESSION['messages'][$key][$type][] = $message;
    }
   backdrop_alter('message', $_SESSION['messages']);
...
}
```
then when we alter:
```
function MODULE_message_alter(&$messages) {
  unset($messages['whatever']; // No problem

  // BUT it would be nice to have $name be available here??
  $messages['whatever'] = t('Thank you @name for your order', ('@name'=> $name)); 
}
```

But you could still _alter with a translated string no? Or am I misinterpreting the problem?

**Comment by jenlampton** [comment]
I've always been frustrated for how messages are handled. They are very hard to alter.  I've updated the issue title to be less about implementation and more about the problem at hand, but I'm all for this. :+1:

**Comment by quicksketch** [comment]
>   // BUT it would be nice to have $name be available here??

Yep, that's the crux of the problem I was describing. Without the substitutions, it would be hard to alter the message effectively.

**Comment by docwilmot** [comment]
There was a Drupal issue for this (postponed x3):
https://www.drupal.org/node/127262

The consensus (though very small discussion) seemed to be to translate the string before passing to `drupal_set_message()`, so no substitution of variables. But I think it would still be better to have the `watchdog()` style and include `$replacements` as well. 

In the end, there are two options; if we could agree on one, we could do it.
- backdrop_message($key, $message, $replacements, $type); // Allows substittions of variables on altering
- backdrop_message($key, $already_translated_message, $type); // simpler, no substitutions.

Finally, 1.x.x or 2.x.x?

**Comment by MrHaroldA** [comment]
I chipped in on the 'format_username everywhere' issue on D.O, since that was not doable by altering the $_SESSION variable because you miss a lot of replacements, like unique one time login url's etc.

If the replacements aren't available, this would not be a "full upgrade" of the current system.
